### PR TITLE
(PUP-9512) Allow discovering and loading of YAML plans

### PR DESCRIPTION
**Pull request by nicklewis**
This adds a second loader for plans which will find plans with a .yaml
extension. Because plans in general, and YAML plans especially, aren't a
meaningful concept in Puppet proper, this loader just calls out to an
injected `yaml_plan_instantiator` to handle actual creation of the plan
once it's been loaded.

This will prefer .pp plans over .yaml when two files with the same name
exist.

### CODE REVIEW

**Comment by nicklewis** [comment]
The corresponding Bolt change is in https://github.com/puppetlabs/bolt/pull/884

**Comment by nicklewis** [comment]
I'm currently looking into how to make `.pp` and `.yaml` versions of the same plan conflict with one another. I expect that involves combining the `PuppetPlanPP` and `PuppetPlanYaml` loaders into a single loader that can operate on multiple files.

**Comment by nicklewis** [comment]
Pushed up a commit that changes the behavior to make it an error to have both a `.pp` and `.yaml` version of the same plan. I had to do some weird things to `module_loaders` to make that happen, not sure if the change makes sense.

File: lib/puppet/pops/loader/puppet_plan_instantiator.rb [hunk 1]
```diff
@@ -9,16 +9,26 @@ class PuppetPlanInstantiator
...
+  def self.create(loader, typed_name, source_refs)
+    if source_refs.length > 1
+      raise ArgumentError, _("Found both a .pp and .yaml version of plan '%{plan_name}' - only one is allowed") % { plan_name: typed_name.name }
+    end
+
+    source_ref = source_refs[0]
+    code_string = Puppet::FileSystem.read(source_ref, :encoding => 'utf-8')
+
+    if source_ref.end_with?('.yaml')
+      return Puppet.lookup(:yaml_plan_instantiator).create(loader, typed_name, source_ref, code_string)
```
Comment (thallgren) [inline lib/puppet/pops/loader/puppet_plan_instantiator.rb]: Shouldn't all added functionality should be dependent on this lookup. There's no need to look for yaml files unless the yaml_plan_instantiator exists.

**Comment by nicklewis** [comment]
@thallgren My [initial implementation](https://github.com/puppetlabs/puppet/pull/7414/commits/c98599b731033dc0153b2e08f47e67e1875d1490) used a separate SmartPath and instantiator and it was definitely much simpler, but it had the drawback that `.pp` files would be preferred over `.yaml` rather than conflicting. Is there a way to make them conflict while having two separate SmartPath instances?

File: lib/puppet/pops/loader/puppet_plan_instantiator.rb (hunk 1 above)
Comment (nicklewis) [reply lib/puppet/pops/loader/puppet_plan_instantiator.rb]: That's a good idea. I'll work on that.

**Comment by nicklewis** [comment]
Pushed another commit to hide all the YAML behavior behind checks for the `yaml_plan_instantiator`.

**Comment by nicklewis** [comment]
@thallgren I think I've address your comments. Is there anything else you want to see on this PR? We're hoping to get it into the next release...

**Pull request merged**